python/dolfinx_contact/unbiased/nitsche_unbiased.py

```diff
@@ -22,22 +21,219 @@
+                        const_coeffs: list[npt.NDArray[np.float64]], consts: npt.NDArray[np.float64]):
+    """
+    Set up newton solver for contact problem.
+    Generate kernels and define functions for updating coefficients, stiffness matrix and residual vector.
+
+    Parameters
+    ==========
+    incr               The load increment in [0,1]
+    F_custom           The compiled form for the residual vector
+    J_custom           The compiled form for the jacobian
+    u                  The displacement from previous step
+    du                 The change in displacement to be computed
+    contact            The contact class
+    markers            The meshtags marking surfaces and domains
+    entities           The contact surface entities for integration
+    quadrature_degre   The quadrature degree
+    const_coeffs       The coefficients for material parameters and h
+    consts             The constants in the forms
+    """
+
+    num_pairs = len(const_coeffs)
+    V = u.function_space
+    mesh = V.mesh
+
+    # generate kernels
+    with common.Timer("~Contact: Generate Jacobian kernel"):
+        kernel_jac = contact.generate_kernel(kt.Jac)
+    with common.Timer("~Contact: Generate residual kernel"):
+        kernel_rhs = contact.generate_kernel(kt.Rhs)
+
+    # create vector and matrix
+    A = contact.create_matrix(J_custom)
+    b = fem.petsc.create_vector(F_custom)
+
+    # Pack gap, normals and test functions on each surface
+    gaps = []
+    normals = []
+    test_fns = []
+    with common.Timer("~Contact: Pack gap, normals, testfunction"):
+        for i in range(num_pairs):
+            gaps.append(contact.pack_gap(i))
+            normals.append(contact.pack_ny(i))
+            test_fns.append(contact.pack_test_functions(i))
+
+    # Concatenate all coeffs
+    ccfs = []
+    for i in range(num_pairs):
+        ccfs.append(np.hstack([const_coeffs[i], gaps[i], normals[i], test_fns[i]]))
+
+    # retrieve boundary conditions for time step
+    tbcs = []
+    for k, g in enumerate(bcs[1]):
+        tag = bcs[0][k][0]
+        sub = bcs[0][k][1]
+        if sub == -1:
+            fn_space = V
+        else:
+            fn_space = V.sub(sub)
+        bdy_dofs = fem.locate_dofs_topological(fn_space, mesh.topology.dim - 1, markers[1].find(tag))
+        tbcs.append(fem.dirichletbc(g, bdy_dofs, fn_space))
+
+    # pack grad u
+    grad_u = []
+    with common.Timer("~~Contact: Pack u"):
+        for i in range(num_pairs):
+            grad_u.append(dolfinx_contact.cpp.pack_gradient_quadrature(
+                u._cpp_object, quadrature_degree, entities[i]))
+
+    # function for updating coefficients coefficients
+    @ common.timed("~Contact: Update coefficients")
+    def compute_coefficients(x, coeffs):
+        size_local = V.dofmap.index_map.size_local
+        bs = V.dofmap.index_map_bs
+        du.x.array[:size_local * bs] = x.array_r[:size_local * bs]
+        du.x.scatter_forward()
```

Comment:
What is the reason for only updating the local bit and then scatter forward, is x not been distributed, or is it because x is a petsc vector?

Reply:
x is a petsc vector I believe